File: clipsync-win/ClipSync/Security/Enrollment.cs
using System;
using System.Security.Cryptography;
using System.Text;

namespace ClipSync.Security;

/// Matches macOS EnrollmentSession: HKDF → session key → HMAC confirm
/// tags → per-direction AES-GCM sub-keys. SPAKE2 group math (RFC 9383
/// edwards25519) is left as a TODO; slot its output secret into
/// Derive().
public sealed class EnrollmentSession
{
    public enum Side { Initiator, Responder }

    public Side Who { get; }
    public string Pin { get; }
    public byte[] OurDid { get; }
    public byte[] TheirDid { get; }
    public byte[] Salt { get; }
    public byte[]? SessionKey { get; private set; }

    public EnrollmentSession(Side who, string pin, byte[] ourDid, byte[] theirDid, byte[] salt)
    {
        Who = who; Pin = pin; OurDid = ourDid; TheirDid = theirDid; Salt = salt;
    }

    public void Derive(byte[] spake2Secret)
    {
        var prk = HMACSHA256.HashData(Salt, spake2Secret);
        SessionKey = HMACSHA256.HashData(prk, Encoding.UTF8.GetBytes("clipsync-enroll-v1"));
    }

    public (byte[] A, byte[] B) ConfirmTags()
    {
        if (SessionKey is null) throw new InvalidOperationException();
        var a = HMACSHA256.HashData(SessionKey, Encoding.UTF8.GetBytes("clipsync-confirm-A"));
        var b = HMACSHA256.HashData(SessionKey, Encoding.UTF8.GetBytes("clipsync-confirm-B"));
        return (a, b);
    }

    public byte[] Seal(byte[] plaintext, string direction)
    {
        if (SessionKey is null) throw new InvalidOperationException();
        var subkey = HMACSHA256.HashData(SessionKey, Encoding.UTF8.GetBytes(direction));
        var nonce = new byte[12];
        var ct = new byte[plaintext.Length];
        var tag = new byte[16];
        using var gcm = new AesGcm(subkey, 16);
        gcm.Encrypt(nonce, plaintext, ct, tag);
        var outBuf = new byte[ct.Length + 16];
        Buffer.BlockCopy(ct, 0, outBuf, 0, ct.Length);
        Buffer.BlockCopy(tag, 0, outBuf, ct.Length, 16);
        return outBuf;
    }

    public byte[] Open(byte[] sealed_, string direction)
    {
        if (SessionKey is null) throw new InvalidOperationException();
        var subkey = HMACSHA256.HashData(SessionKey, Encoding.UTF8.GetBytes(direction));
        var nonce = new byte[12];
        var ct = new byte[sealed_.Length - 16];
        var tag = new byte[16];
        Buffer.BlockCopy(sealed_, 0, ct, 0, ct.Length);
        Buffer.BlockCopy(sealed_, ct.Length, tag, 0, 16);
        var pt = new byte[ct.Length];
        using var gcm = new AesGcm(subkey, 16);
        gcm.Decrypt(nonce, ct, tag, pt);
        return pt;
    }
}

// TODO: port an RFC 9383 SPAKE2 over edwards25519 implementation (e.g.
// from the Matter/CHIP C++ reference or Go x/crypto). Pipe its shared
// secret into EnrollmentSession.Derive().

File: clipsync-mac/Sources/ClipSync/Security/Enrollment.swift
import Foundation
import Crypto

/// Enrollment is the PIN-gated, one-shot identity exchange described in
/// PROTOCOL.md §7. It runs on a *raw* TCP connection (no TLS) because
/// the pair doesn't yet trust each other's certs — SPAKE2 itself is
/// what makes the exchange safe.
///
/// This file stubs the SPAKE2 state machine interface and derives the
/// confirm tags / session key; the actual SPAKE2 group math lives in
/// `SPAKE2Ed25519.swift` (see TODO at bottom) and is omitted from this
/// scaffold — swift-crypto does not ship SPAKE2 yet, so a small vendored
/// RFC 9383 implementation is needed. The structure here is what the UI
/// calls into once that lands.

struct EnrollmentSession {
    enum Side { case initiator, responder }
    let side: Side
    let pin: String
    let ourDid: Data
    let theirDid: Data
    let salt: Data
    private(set) var sessionKey: Data?

    mutating func derive(from spake2Secret: Data) {
        // HKDF-SHA256(spake2Secret, salt, "clipsync-enroll-v1", 32)
        let prk = HMAC<SHA256>.authenticationCode(
            for: spake2Secret, using: SymmetricKey(data: salt))
        let key = HMAC<SHA256>.authenticationCode(
            for: Data("clipsync-enroll-v1".utf8), using: SymmetricKey(data: Data(prk)))
        self.sessionKey = Data(key)
    }

    func confirmTags() -> (a: Data, b: Data)? {
        guard let k = sessionKey else { return nil }
        let key = SymmetricKey(data: k)
        let a = HMAC<SHA256>.authenticationCode(for: Data("clipsync-confirm-A".utf8), using: key)
        let b = HMAC<SHA256>.authenticationCode(for: Data("clipsync-confirm-B".utf8), using: key)
        return (Data(a), Data(b))
    }

    /// AES-256-GCM seal/open with a zero nonce (safe: the key is used
    /// for at most one message per direction, and directions have
    /// distinct sub-keys derived from HKDF info below).
    func seal(_ plaintext: Data, direction: String) throws -> Data {
        guard let k = sessionKey else { throw CodecError.invalid }
        let subkey = HMAC<SHA256>.authenticationCode(
            for: Data(direction.utf8), using: SymmetricKey(data: k))
        let nonce = try AES.GCM.Nonce(data: Data(repeating: 0, count: 12))
        let box = try AES.GCM.seal(plaintext, using: SymmetricKey(data: Data(subkey)), nonce: nonce)
        return box.ciphertext + box.tag
    }

    func open(_ sealed: Data, direction: String) throws -> Data {
        guard let k = sessionKey, sealed.count >= 16 else { throw CodecError.invalid }
        let subkey = HMAC<SHA256>.authenticationCode(
            for: Data(direction.utf8), using: SymmetricKey(data: k))
        let nonce = try AES.GCM.Nonce(data: Data(repeating: 0, count: 12))
        let ct = sealed.prefix(sealed.count - 16)
        let tag = sealed.suffix(16)
        let box = try AES.GCM.SealedBox(nonce: nonce, ciphertext: ct, tag: tag)
        return try AES.GCM.open(box, using: SymmetricKey(data: Data(subkey)))
    }
}

// TODO: vendor an RFC 9383 SPAKE2 over edwards25519 implementation. Feed
// its output secret into `EnrollmentSession.derive(from:)`. Swift-crypto
// does not currently ship SPAKE2; see e.g. the Matter/CHIP reference
// implementation or go-spake2 for a clean port.
